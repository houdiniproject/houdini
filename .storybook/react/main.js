// License: LGPL-3.0-or-later
// Based on https://github.com/rails/webpacker/issues/1004#issuecomment-628377930
process.env.NODE_ENV = "development";
const railsWebpackEnv = require("../../config/webpack/environment");

module.exports = {
  stories: [
    "../../app/packs/**/*!(--html).stories.mdx",
    "../../app/packs/**/*!(--html).stories.[tj]s?(x)"
  ],
  addons: [
    "@storybook/addon-links",
    "@storybook/addon-essentials",
    'storybook-addon-intl'
  ],
  webpackFinal: (config) => {
    const result = {
    // do mutation to the config
    ...config,
    resolve: {
      ...config.resolve,
      ...railsWebpackEnv.config.resolve,
      modules: railsWebpackEnv.resolvedModules.map((i) => i.value),
    },
    module: {
      ...config.module,
      rules: railsWebpackEnv.loaders
        .filter((i) => !["nodeModules", //not sure why this is here
        "moduleCss" // this addresses issues with our webpack config for css not matching what storybook wants
      ].includes(i.key) )
        .map((i) => i.value),
    },
    plugins: [
      ...config.plugins,
      ...railsWebpackEnv.plugins.map((i) => i.value),
    ],
    
  };

  result.module.rules.filter((i) => i.test.test('.ttf')).forEach((i) => {
    i.use=['url-loader']
  })


  return result;
},
};
