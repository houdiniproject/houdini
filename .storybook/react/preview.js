// License: LGPL-3.0-or-later
import intlDecorate from '../../app/packs/components/tests/intl';
const jest = require('jest-mock');
window.jest = jest;

export const decorators = [intlDecorate()]