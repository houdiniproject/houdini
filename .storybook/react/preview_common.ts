// License: LGPL-3.0-or-or-later

import intlDecorate from '../../app/javascript/components/tests/intl';
import cssBaseline from '../../app/javascript/components/tests/decorators/baseline'
import clearSessionStorage from '../../app/javascript/components/tests/decorators/clearSessionStorage'
import { initialize, mswDecorator } from 'msw-storybook-addon';

initialize({
	onUnhandledRequest: 'bypass' // we need this because otherwise HMR and other parts of Storybook loading wouldn't work
});


export const decorators = [ clearSessionStorage(), mswDecorator, intlDecorate(), cssBaseline()]