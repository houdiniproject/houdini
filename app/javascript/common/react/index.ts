// License: LGPL-3.0-or-later

/**
 * Adds type safety to your default props.
 *
 * @example
 * // your component
 * function CustomComponent(props:{className:string, disabled?:boolean}) : JSX.Element {
 * 	// not show for
 * }
 *
 * CustomComponent.defaultProps = defaultProps(CustomComponent, {
 * 	className: 'another'
 * })
 *
 * @example <caption>Function properly typechecks</caption>
 * // use the same component as previous example
 * CustomComponent.defaultProps = defaultProps(CustomComponent, {
 * 	className: 'another',
 * 	unexpectedProp: 3 // <-- Typecheck will throw a compile error because you didn't expect this.
 * })
 *
 * @param {(props:TProperty) => JSX.Element} _component - the component you want to create default props for
 * @param {Partial<TProperty>} defaultProps - the defaultProps for your component
 * @returns defaultProps
 * */
export function defaultProps<TProperty>(_component:(props:TProperty) => JSX.Element, defaultProps:Partial<TProperty>):Partial<TProperty> {
	return defaultProps;
}