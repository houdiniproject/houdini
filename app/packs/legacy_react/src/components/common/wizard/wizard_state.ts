// License: LGPL-3.0-or-later
import {observable, action, computed, toJS, reaction, runInAction} from "mobx";
import {Field, Form, FieldDefinition, FieldHandlers, FieldHooks} from "mobx-react-form";
import * as _ from "lodash";
import {AbstractWizardState, AbstractWizardTabPanelState} from "./abstract_wizard_state";

interface SubFormDefinition {
  related?: string[]
  bindings?: any
  options?: any
  extra?: any
  hooks?: FieldHooks
  handlers?: FieldHandlers
  fields?: Array<FieldDefinition>
}

export abstract class WizardState<PanelStateType extends WizardTabPanelState = WizardTabPanelState,
  FormStateType extends Form = Form> extends AbstractWizardState<PanelStateType> {

  protected constructor(panelType: { new(): PanelStateType }) {
    super(panelType)
  }

  @observable form: FormStateType;

  abstract createForm(i: any): FormStateType;

  addTab(tab: { tabName: string, label: string, tabFieldDefinition: SubFormDefinition}): PanelStateType {
    const ret = super.addTab(tab);

    runInAction(() => {
      ret.panelFormDefinition = tab.tabFieldDefinition as FieldDefinition
    });

    return ret;
  }

  @action.bound
  initialize(): void {
    if (this.panels.length > 0) {
      //let's create the forms
      const lastIndex = this.panels.length;
      for (let i = 0; i < lastIndex; i++) {
        let ourPanel = this.panels[i]

        if (!ourPanel.panelFormDefinition.hooks)
          ourPanel.panelFormDefinition.hooks = {}

        //ourPanel.originalOnSuccessHook = toJS(ourPanel.panelFormDefinition.hooks['onSuccess'])

        ourPanel.panelFormDefinition.hooks['onSuccess'] = this.onSuccessForPanel

        /// this won't work because the hook is already replaced
        // if (ourPanel.panelFormDefinition.hooks)
        //   ourPanel.originalOnErrorHook = ourPanel.panelFormDefinition.hooks['onError']
        ourPanel.panelFormDefinition.hooks['onError'] = this.onErrorForPanel

        ourPanel.panelFormDefinition.name = ourPanel.tabName
      }

      //we need to change these back to JS objects because they're likely observable and fieldDefinitions
      // can't handle that
      const fieldDefinition = toJS(this.panels.map((i) => toJS(i.panelFormDefinition)))
      this.form = this.createForm({fields: fieldDefinition})

      _.forEach(this.panels, (i) => {
        //add the form to each panel
        i.parentForm = this.form
        i.form = this.form.$(i.tabName)
      })
    }
  }

  @action.bound
  onSuccessForPanel(a: Field): void {

    // if (this.activeTab.originalOnSuccessHook) {
    //   this.activeTab.originalOnSuccessHook(a)
    // }

    if (a.submitting) {
      if (this.nextTab)
        this.moveToNextTab();
      else
        this.form.submit();
    }
  }


  @action.bound
  onErrorForPanel(a: Field): any {
    // if (this.activeTab.originalOnErrorHook) {
    //   this.activeTab.originalOnErrorHook(a)
    // }
  }
}

export class WizardTabPanelState<ParentFormStateType extends Form = Form> extends AbstractWizardTabPanelState {
  @observable parentForm: ParentFormStateType

  @observable form: Field


  // @observable originalOnSuccessHook: Function
  // @observable originalOnErrorHook: Function

  panelFormDefinition: FieldDefinition

  /**
   * Whether this tab's form is valid. We override this in a mock so we can manually set the validity
   * via a simple function call
   * @returns {boolean} true if this tab's form is valid, otherwise false
   */
  @computed
  get isValid(): boolean {
    return this.form.isValid
  }
}