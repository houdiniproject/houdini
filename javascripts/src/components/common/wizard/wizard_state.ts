// License: LGPL-3.0-or-later
import {observable, action, computed, toJS, reaction, runInAction} from "mobx";
import {Field, Form, FieldDefinition, FieldHandlers, FieldHooks} from "mobx-react-form";
import _ = require("lodash");
import {TabManager} from "./manager"
import {Wizard} from "./Wizard";

interface SubFormDefinition {
  related?: string[]
  bindings?: any
  options?: any
  extra?: any
  hooks?: FieldHooks
  handlers?: FieldHandlers
  fields?: Array<FieldDefinition>
}

export abstract class WizardState<PanelStateType extends WizardTabPanelState = WizardTabPanelState> {
  panelType: { new(): PanelStateType }

  constructor(panelType: { new(): PanelStateType } = null) {
    this.panelType = panelType
  }

  @observable
  lastRequestedTab: WizardTabPanelState

  @observable panels = new Array<WizardTabPanelState>()
  @observable form: Form

  @observable manager: TabManager

  abstract createForm(i: any): Form;

  @action.bound
  private createChildState(): WizardTabPanelState {
    if (this.panelType)
      return new this.panelType()
    else
      return new WizardTabPanelState()
  }

  @action.bound
  addTab(tabName: string, label: string, tabFieldDefinition: SubFormDefinition): WizardTabPanelState {

    var newTab = this.createChildState()
    newTab.id = _.uniqueId('tab')
    newTab.tabName = tabName
    newTab.label = label
    if (this.panels.length == 0) {
      this.activeTab = newTab
    }
    newTab.parent = this
    newTab.panelFormDefinition = tabFieldDefinition as FieldDefinition
    this.panels.push(newTab)

    if (!this.manager) {
      this.manager = new TabManager({
        onChange: this.handleTabChange, letterNavigation: true, activeTabId: this.activeTab.id,
        canChangeTo: this.canChangeTo
      })
    }


    reaction(() => this.lastConsistentlyEnabledTab, (data, react) => {
      if (data.before(this.activeTab)) {
        this.activateTab(data)
      }
    })
    return newTab;
  }

  @action.bound
  initialize(): void {
    if (this.panels.length > 0) {
      //let's create the forms
      let lastIndex = this.panels.length
      for (let i = 0; i < lastIndex; i++) {
        let ourPanel = this.panels[i]

        if (!ourPanel.panelFormDefinition.hooks)
          ourPanel.panelFormDefinition.hooks = {}

        ourPanel.originalOnSuccessHook = toJS(ourPanel.panelFormDefinition.hooks['onSuccess'])

        ourPanel.panelFormDefinition.hooks['onSuccess'] = this.onSuccessForPanel

        /// this won't work because the hook is already replaced
        if (ourPanel.panelFormDefinition.hooks)
          ourPanel.originalOnErrorHook = ourPanel.panelFormDefinition.hooks['onError']
        ourPanel.panelFormDefinition.hooks['onError'] = this.onErrorForPanel

        ourPanel.panelFormDefinition.name = ourPanel.tabName
      }

      //we need to change these back to JS objects because they're likely observable and fieldDefinitions
      // can't handle that
      let fieldDefinition = toJS(this.panels.map((i) => toJS(i.panelFormDefinition)))
      this.form = this.createForm({fields: fieldDefinition})

      _.forEach(this.panels, (i) => {
        //add the form to each panel
        i.parentForm = this.form
        i.form = this.form.$(i.tabName)
      })
    }
  }

  @computed
  get tabsByName(): { [name: string]: WizardTabPanelState } {
    return _.fromPairs(this.panels.map((i) => [i.tabName, i]));
  }

  @observable activeTab: WizardTabPanelState

  activateTab(tab: WizardTabPanelState | string) {
    let tabId: string = null
    if (tab instanceof WizardTabPanelState) {
      tabId = tab.id
    }
    else {
      tabId = tab;
    }

    this.manager.activateTab(tabId)
  }

  @action.bound
  handleTabChange(tabId: string): WizardTabPanelState {
    let self = this

    let tabInfo = _.find(self.panels, (i) => i.id == tabId)
    if (tabInfo && tabInfo.enabled) {
      this.activeTab = tabInfo

      return self.activeTab === tabInfo ? tabInfo : null;
    }
    return null
  }

  @action.bound
  private canChangeTo(tabId: string): boolean {

    let tab = _.find(this.panels, (i) => i.id == tabId)
    return tab && tab.enabled

  }

  @action.bound
  moveToNextTab() {
    let self = this

    if (this.nextTab) {
      self.manager.activateTab(this.nextTab.id)

    }
  }

  @action.bound
  onSuccessForPanel(a: Field): void {

    if (this.activeTab.originalOnSuccessHook) {
      this.activeTab.originalOnSuccessHook(a)
    }

    if (a.submitting) {
      if (this.nextTab)
        this.moveToNextTab()
      else
        this.form.submit()
    }
  }


  @action.bound
  onErrorForPanel(a: Field): any {
    if (this.activeTab.originalOnErrorHook) {
      this.activeTab.originalOnErrorHook(a)
    }
  }

  @computed
  get firstDisabledTab(): WizardTabPanelState {
    return _.find(this.panels, (i) => !i.enabled)
  }

  @computed
  get lastConsistentlyEnabledTab(): WizardTabPanelState {
    return this.firstDisabledTab ? this.firstDisabledTab.previous : _.last(this.panels)
  }


  @computed
  get nextTab(): WizardTabPanelState {
    return this.activeTab.next
  }

  @computed
  get previousTab(): WizardTabPanelState {
    return this.activeTab.previous
  }
}

export class WizardTabPanelState {
  @observable parent: WizardState

  @observable parentForm: Form

  @observable form: Field


  @observable id: string
  @observable tabName: string
  @observable label: string

  @observable originalOnSuccessHook: Function
  @observable originalOnErrorHook: Function

  panelFormDefinition: FieldDefinition

  /**
   * Whether this tab's form is valid. We override this in a mock so we can manually set the validity
   * via a simple function call
   * @returns {boolean} true if this tab's form is valid, otherwise false
   */
  @computed
  get isValid(): boolean {
    return this.form.isValid
  }

  @computed
  get active(): boolean {
    return this.parent.activeTab === this
  }

  @computed
  get enabled(): boolean {

    let previous = this.previous
    let next = this.next

    if (previous) {
      let enabled = previous.enabled
      let valid = previous.isValid;
      return enabled && valid
    }
    return true
  }


  @computed
  get previous(): WizardTabPanelState {
    if (!this.parent || !this.parent.panels)
      return null;
    let index = _.findIndex(this.parent.panels, (i) => i == this)
    if (index === null) {
      // return null but we have a problem here
      return null
    }
    if (index === 0) {
      // there is no previous one because we're first!
      return null;
    }

    return this.parent.panels[index - 1]
  }

  @computed
  get next(): WizardTabPanelState {
    if (!this.parent || !this.parent.panels)
      return null;

    let index = _.findIndex(this.parent.panels, (i) => i == this)
    let panelLength = this.parent.panels.length
    if (index === null) {
      // return null but we have a problem here
      return null
    }

    if (index + 1 >= panelLength) {
      //we have no advanced
      return null;
    }

    return this.parent.panels[index + 1]
  }

  before(tab: WizardTabPanelState): boolean {
    let testItem: WizardTabPanelState = this
    while (testItem.next != tab) {
      if (!testItem.next)
        return false;
      testItem = testItem.next
    }

    return true
  }


  after(tab: WizardTabPanelState): boolean {
    let testItem: WizardTabPanelState = this
    while (testItem.previous != tab) {
      if (!testItem.previous)
        return false;
      testItem = testItem.previous
    }

    return true
  }
}