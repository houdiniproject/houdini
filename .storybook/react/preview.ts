// License: LGPL-3.0-or-later
import {decorators} from './preview_common';

require('../../app/javascript/components/tests/decorators/roboto')

export {decorators};
