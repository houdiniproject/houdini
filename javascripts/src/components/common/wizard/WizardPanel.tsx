// License: LGPL-3.0-or-later
import * as React from 'react';
import { observer } from 'mobx-react'
import { WizardTabPanelState} from './wizard_state';
import {computed} from 'mobx';
import * as _ from 'lodash'
import {TabPanel} from "./RAT/TabPanel";


export interface WizardTabPanelProps {
  tab: WizardTabPanelState
}

export interface WizardPanelProps extends WizardTabPanelProps {
    [props:string]:any
}

@observer
export class WizardPanel extends React.Component<WizardPanelProps, {}> {
    @computed
    get tab():WizardTabPanelState{
        return this.props.tab
    }
    @computed
    get isActive(){
        return this.tab.active
    }
    render() {

        let props = _.omit(this.props, ['tab'])
        return <TabPanel {...props} tabId={this.tab.id} active={this.isActive}
         className="wizard-step">
          {this.props.children}
        </TabPanel>
    }
}

