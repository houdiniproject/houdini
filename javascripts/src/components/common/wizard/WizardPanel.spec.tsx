// License: LGPL-3.0-or-later
import * as React from 'react';
import 'jest';

import * as Component from './WizardPanel'

import {Form, Field} from 'mobx-react-form';
import { shallow, render } from 'enzyme';
import {TabPanel, Wrapper} from 'react-aria-tabpanel'
import toJson from 'enzyme-to-json';
import {WizardState} from "./wizard_state";
const TestRenderer = require('react-test-renderer')

class EasyWizardState extends WizardState{

  createForm(i: any): Form {
    return new Form(i)
  }

}

describe('WizardPanel', () => {
    test('shallow render', () => {
      let fields = [{name: 'fun', id: 'fun', label: 'alsofun'}]
      const form = new Form({fields});

      const ws = new EasyWizardState()
      ws.addTab('something', 'something label',{} )
      ws.initialize()

      const tree = shallow(<Component.WizardPanel tab={ws.tabsByName['something']} anotherProp={false}><hr/></Component.WizardPanel>)
        
        
      expect(toJson(tree)).toMatchSnapshot()
    })
})