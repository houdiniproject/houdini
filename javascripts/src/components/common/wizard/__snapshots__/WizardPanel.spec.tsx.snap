// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardPanel shallow render 1`] = `
<AriaTabPanel-TabPanel
  active={true}
  anotherProp={false}
  className="wizard-step"
  tabId="tab1"
  tag="div"
>
  <hr />
</AriaTabPanel-TabPanel>
`;
